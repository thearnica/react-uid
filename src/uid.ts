/**
 * generates a UID factory
 * @internal
 * @example
 * const uid = generateUID();
 * uid(object) = 1;
 * uid(object) = 1;
 * uid(anotherObject) = 2;
 */
export const generateUID = () => {
  let counter = 1;

  const map = new WeakMap<any, number>();

  /**
   * @borrows {uid}
   */
  const uid = (item: any, index?: number): string => {
    if (typeof item === 'number' || typeof item === 'string') {
      return index ? `idx-${index}` : `val-${item}`;
    }

    if (!map.has(item)) {
      map.set(item, counter++);

      return uid(item);
    }

    return 'uid' + map.get(item);
  };

  return uid;
};

/**
 * @name uid
 * returns an UID associated with {item}
 * @param {Object} item - object to generate UID for
 * @param {Number} index, a fallback index
 * @example
 * uid(object) == 1;
 * uid(object) == 1;
 * uid(anotherObject) == 2;
 * uid("not object", 42) == 42
 *
 * @see {@link useUID}
 */
export const uid = generateUID();
