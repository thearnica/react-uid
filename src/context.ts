import * as React from 'react';

import { generateUID } from './uid';

export type IdSourceType = {
  value: number;
  prefix: string;
  uid: (item: any, index?: number) => string;
};

export const createSource = (prefix = ''): IdSourceType => ({
  value: 1,
  prefix: prefix,
  uid: generateUID(),
});

export interface UIDProps {
  name?: (n: string | number) => string;
  idSource?: IdSourceType;
  children: (id: string, uid: (item: any, index?: number) => string) => React.ReactNode;
}

export const counter = createSource();

export const source = React.createContext(createSource());

export const getId = (source: IdSourceType): number => source.value++;
export const getPrefix = (source?: IdSourceType): string => (source ? source.prefix : '');
