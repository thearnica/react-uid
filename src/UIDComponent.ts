import * as React from 'react';

import { counter, getId, getPrefix, UIDProps } from './context';

// --------------------------------------------

const prefixId = (id: number | string, prefix: string, name: UIDProps['name']) => {
  const uid = prefix + id;

  return String(name ? name(uid) : uid);
};

/**
 * @deprecated
 * UID in form of renderProps (not SSR friendly)
 * @see https://github.com/thearnica/react-uid#react-components
 * @example
 * // get UID to connect label to input
 * <UID>
 *   {(id)} => <label htmlFor={id}><input id={id}/>}
 * </UID>
 *
 * // get uid to generate uid for a keys in a list
 * <UID>
 *   {(, uid)} => items.map(item => <li key={uid(item) />)}
 * </UID>
 */
export class UID extends React.Component<UIDProps> {
  state = {
    quartz: this.props.idSource || counter,
    prefix: getPrefix(this.props.idSource),
    id: getId(this.props.idSource || counter),
  };

  uid = (item: any): string =>
    prefixId(this.state.id + '-' + this.state.quartz.uid(item), this.state.prefix, this.props.name);

  render() {
    const { children, name } = this.props;
    const { id, prefix } = this.state;

    return children(prefixId(id, prefix, name), this.uid);
  }
}
