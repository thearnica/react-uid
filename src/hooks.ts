import * as React from 'react';

import { useContext, useState } from 'react';

import { counter, getId, getPrefix, IdSourceType, source } from './context';

/**
 * Actually a uid function - call with any argument
 * @returns {String}
 * @see {@link uid}
 */
type SeedGenerator = (id: any) => string;
type GeneratedUID = {
  uid: string;
  gen(item: any): string;
};

const generateUID = (context: IdSourceType): GeneratedUID => {
  const quartz = context || counter;
  const prefix = getPrefix(quartz);
  const id = getId(quartz);
  const uid = prefix + id;
  const gen = (item: any) => uid + quartz.uid(item);

  return { uid, gen };
};

const useUIDState = (): GeneratedUID => {
  if (process.env.NODE_ENV !== 'production') {
    if (!('useContext' in React)) {
      throw new Error('Hooks API requires React 16.8+');
    }
  }

  const context = useContext(source);
  const [uid] = useState<GeneratedUID>(() => generateUID(context));

  return uid;
};

/**
 * returns and unique id. SSR friendly
 * returns {String}
 * @see {@link UIDConsumer}
 * @see https://github.com/thearnica/react-uid#hooks-168
 * @example
 * const id = useUID();
 * id == 1; // for example
 */
export const useUID = (): string => {
  const { uid } = useUIDState();

  return uid;
};

/**
 * returns an uid generator
 * @see {@link UIDConsumer}
 * @see https://github.com/thearnica/react-uid#hooks-168
 * @example
 * const uid = useUIDSeed();
 * return (
 *  <>
 *    <label for={seed('email')}>Email: </label>
 *    <input id={seed('email')} name="email" />
 *    {data.map(item => <div key={seed(item)}>...</div>
 *  </>
 * )
 */
export const useUIDSeed = (): SeedGenerator => {
  const { gen } = useUIDState();

  return gen;
};
