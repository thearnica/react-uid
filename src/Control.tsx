import * as React from 'react';
import {createSource, UIDProps, source} from "./context";
import {UID} from "./UIDComponent";

interface WithPrefix {
  prefix?: string;
}

/**
 * UID isolation component, required for SSR and testing.
 * Wrap your application with it to guarantee UID consistency between SSR and CSR.
 * @param {String} [prefix] - prefix for all generated ids
 * @example
 * <UIDReset>
 *    <App />
 * </UIDReset/>
 * @see https://github.com/thearnica/react-uid#server-side-friendly-uid
 */
export const UIDReset: React.FC<React.PropsWithChildren<WithPrefix>> = ({children, prefix = ''}) => (
  <source.Provider value={createSource(prefix)}>{children}</source.Provider>
);

/**
 * Creates a sub-ids for nested components, isolating from inside a branch.
 * Useful for self-contained elements or code splitting
 * @see https://github.com/thearnica/react-uid#code-splitting
 */
export const UIDFork: React.FC<React.PropsWithChildren<WithPrefix>> = ({children, prefix = ''}) => (
  <UIDConsumer>
    {(id) => (
      <source.Provider value={createSource(id + '-' + prefix)}>
        {children}
      </source.Provider>
    )}
  </UIDConsumer>
);

/**
 * UID in form of renderProps. Supports nesting and SSR. Prefer {@link useUID} hook version if possible.
 * @see https://github.com/thearnica/react-uid#server-side-friendly-uid
 * @see https://github.com/thearnica/react-uid#react-components
 * @example
 * // get UID to connect label to input
 * <UIDConsumer>
 *   {(id)} => <label htmlFor={id}><input id={id}/>}
 * </UIDConsumer>
 *
 * // get uid to generate uid for a keys in a list
 * <UIDConsumer>
 *   {(, uid)} => items.map(item => <li key={uid(item) />)}
 * </UIDConsumer>
 *
 * @see {@link useUID} - a hook version of this component
 * @see {@link UID} - not SSR compatible version
 */
export const UIDConsumer: React.FC<UIDProps> = ({name, children}) => (
  <source.Consumer>
    {(value) => (
      <UID name={name} idSource={value} children={children}/>
    )}
  </source.Consumer>
);