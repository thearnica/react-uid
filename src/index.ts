import { UIDReset, UIDConsumer, UIDFork } from './Control';
import { UID } from './UIDComponent';
import { useUID, useUIDSeed } from './hooks';
import { generateUID, uid } from './uid';

export { generateUID, uid, UID, UIDConsumer, UIDReset, UIDFork, useUID, useUIDSeed };
