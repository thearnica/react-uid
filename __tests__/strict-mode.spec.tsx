import { render } from '@testing-library/react';
import * as React from 'react';

import { UIDReset, useUID } from '../src';

describe('uid hooks in strict mode', () => {
  it('test uid', () => {
    const Target = ({ x }: any) => {
      const id = useUID();

      return (
        <i>
          {x} == {id} ?
        </i>
      );
    };

    const markup = (
      <React.StrictMode>
        <UIDReset prefix={''}>
          <Target x={1} />
        </UIDReset>
      </React.StrictMode>
    );

    const test = render(markup);
    expect(test.baseElement.innerHTML).toMatchSnapshot();
  });
});
