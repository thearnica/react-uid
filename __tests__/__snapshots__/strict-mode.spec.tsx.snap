// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uid hooks in strict mode test uid 1`] = `"<div><i>1 == 2 ?</i></div>"`;
